modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,273 @@
+                    break;
+            }
+        }
+    }
+
+    /**
+     * Obtain an Authorization Code
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function authenticate(&$state)
+    {
+        // We are going to need the authId in order to retrieve this authentication source later
+        $state[self::AUTHID] = $this->authId;
+
+        $stateID = SimpleSAML_Auth_State::getStateId($state);
+        SimpleSAML_Logger::debug("globus auth at state = " . $stateID);
+
+        $consumer = new sspmod_oauth_Consumer($this->key, $this->secret);
+
+        $request = array(
+            'client_id' => $this->key,
+            'scope' => $this->scope,
+            'state' => $stateID,
+            'response_type' => 'code',
+            'redirect_uri' => $this->redirect_uri
+        );
+
+        $state['globusauth:request'] = $request;
+        $urlAppend = \SimpleSAML\Utils\HTTP::addURLParameters(self::API_ENDPT . '/oauth2/authorize', $request);
+        SimpleSAML_Auth_State::saveState($state, self::STAGE_INIT);
+
+        // Get an Authorization Code from Globus
+        /* Example: https://auth.globus.org/v2/oauth2/authorize?client_id=69ba5e62-7285-45db-952d-e0bb73b5eac7&scope=urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:auth.globus.org:view_identities offline_access&response_type=code&redirect_uri=https://www.example.org/my_app/login&state=g6l14b2xlgx4dtce8d2ja714i
+        */
+        $authorizeUrl = $consumer->getAuthorizeRequest($urlAppend, $request);
+    }
+
+    /**
+     * Exchange authorization code for an access token
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function finalStep(&$state)
+    {
+        $request = array(
+            'code' => $_GET['code'],
+            'redirect_uri' => $this->redirect_uri,
+            'grant_type' => 'authorization_code'
+        );
+        $oauthSignature = $this->key . ":" . $this->secret;
+        
+        // Exchange the code we got earlier for an access token
+        $result = $this->doCurlRequest('/oauth2/token', $oauthSignature, null, null, $request, 'POST');
+        $otherTokens = $result->other_tokens;
+
+        // Use this access token to tell us about our current user + affiliations
+        $userInfo = $this->doCurlRequest('/oauth2/userinfo', null, $result->access_token);
+
+        // Get any other identities we may have
+        $otherIdentities = array();
+
+        if (isset($otherTokens)) {
+            foreach ($otherTokens as &$item) {
+                $res = $this->doCurlRequest('/api/identities/', null, $item->access_token, $userInfo['sub']);
+
+                foreach ($res as &$id) {
+                    array_push($otherIdentities, $id);
+                }
+            }
+        }
+
+        $otherIdentities = $this->flatten($otherIdentities, 'identity.');
+        $fullname = $this->getFullName($userInfo['name']);
```

Comment:
are we sure this is pulling the name from the primary identity?
Should we be able to say get the name from identity X?

From the documentation it seems like you can have name in each linked identity, so what one is it actually getting?

Reply:
The uppermost one, not a nested one. The object looks something like:
`{
name : value,
fieldX: value,
fieldY: value,
identities: [
name: value,
fieldX: value,
etc..
]
}`

Reply:
So that is my question, is that the globus one or say the UBIT one, Should we be pulling a specific one should we have it as something that can be customized?

For XSEDE should we use the XSEDE Identity?

Reply:
Probably something we need to discuss next week + get a consensus on.